### PR TITLE
[WIP][Test] reenable shutdown tests for net_test

### LINKED ISSUE #944: Reenable shutdown tests for net_test.ts
Currently, tests are disabled for `netCloseReadSuccess`, `netDoubleCloseRead`, `netCloseWriteSuccess`, `netDoubleCloseWrite`. The tests always pass on my local machine (macOS), but it seems the behavior is different on other platforms. (I myself don't have a windows machine)

Add this issue to avoid forgetting about it & in case others would like to contribute and fix them

### PR DESCRIPTION
**Pull request by sasurau4**
## Overview
fixes #944

### CODE REVIEW

File: js/net_test.ts [hunk 1]
```diff
@@ -193,24 +195,22 @@ testPerm({ net: true }, async function netCloseWriteSuccess() {
     err = e;
   }
   assert(!!err);
-  assertEquals(err.kind, Deno.ErrorKind.BrokenPipe);
-  assertEquals(err.name, "BrokenPipe");
+  assertEquals(err, new Error("write error"));
```
Comment (sasurau4) [inline js/net_test.ts]: Is it better to handle such error as DenoError?
https://github.com/denoland/deno/blob/master/js/files.ts#L163

Comment (ry) [reply js/net_test.ts]: Yeah

Looks like you're failing on Windows. I forget what the deal with this test was. Maybe just flaky. Try running it couple of times locally.

Comment (sasurau4) [reply js/net_test.ts]: Ok

I found the cause of failing on Windows is same to #2358. 
I'll try to deal with it.

**Comment by sasurau4** [comment]
@ry 
I found https://github.com/hyperium/hyper/issues/1575#issuecomment-399161972
It seems the cause of test failing on Windows. 
So, I want to try using `net2` crate and `tokio::net::TcpListener::from_std()` function instead of `TcpListener::bind()`.
https://docs.rs/tokio/0.1.20/tokio/net/struct.TcpListener.html#method.from_std
I need your help how to add `net2` crate to thrid_party dir.
Thanks in advance.

**Comment by sasurau4** [comment]
📝 
At https://github.com/denoland/deno/pull/2410/commits/4719f576b912ddbd603a7cc58aa056b4a27eef94, the net test failed at `netCloseReadSuccess` by `BrokenPipe: read error` and others by `AddrInUse: Only one usage of each socket address (protocol/network address/port) is normally permitted. (os error 10048)`.

At https://github.com/denoland/deno/pull/2410/commits/e5a1438ce1d0b30239267316e5e42eaf736b5554, the net test failed at `netCloseReadSuccess` by `BrokenPipe`.

So, the difference of 4719f57 and e5a1438 is enable `reuse_address`.

I couldn't find why the test failed by `BrokenPipe: read error`. 🤔

**Comment by NamPNQ** [comment]
@sasurau4 
For case `netCloseReadSuccess`, and another should do samething, you can follow as
```ts
async function netCloseReadSuccess() {
    const addr = "127.0.0.1:4500";
    const listener = Deno.listen("tcp", addr);
    const closeDeferred = deferred();
    const closeReadDeferred = deferred();
    listener.accept().then(async conn => {
      await closeReadDeferred;
      //await conn.write(new Uint8Array([1, 2, 3])); (*1)
      const buf = new Uint8Array(1024);
      const readResult = await conn.read(buf);
      assertEquals(3, readResult.nread);
      assertEquals(4, buf[0]);
      assertEquals(5, buf[1]);
      assertEquals(6, buf[2]);
      conn.close();
      closeDeferred.resolve();
    });
    const conn = await Deno.dial("tcp", addr);
   conn.closeRead(); // closing read
    closeReadDeferred.resolve();
    const buf = new Uint8Array(1024);
    //const readResult = await conn.read(buf); (*2)
    //assertEquals(0, readResult.nread); // No error, read nothing
    //assertEquals(true, readResult.eof); // with immediate EOF
    // Ensure closeRead does not impact write
    await conn.write(new Uint8Array([4, 5, 6]));
    await closeDeferred;
    listener.close();
    conn.close();
  }
```

Notes:
(*1, *2): You can improve it to raise error 10058, follow link: https://docs.microsoft.com/en-us/windows/desktop/winsock/windows-sockets-error-codes-2

**Comment by NamPNQ** [comment]
Something you can do for improve behavior is:

Add `fn raw_os_error` in `cli/errors.rs`
```
pub fn raw_os_error(&self) -> Option<i32> {
        match self.repr {
            Repr::IoErr(ref err) => err.raw_os_error(),
            _ => None,
        }
}
```
Edit `cli/dispatch_minimal.rs` look like
```
Err(err) => {
        // TODO(ry) The dispatch_minimal doesn't properly pipe errors back to
        // the caller.
        debug!("swallowed err {}", err);
        record.result = match err.raw_os_error() {
          None => -1,
          Some(i) => i * -1
        };
 }
```
And handle better error in `js/files.ts`, for example
```
export async function read(rid: number, p: Uint8Array): Promise<ReadResult> {
  const nread = await sendAsyncMinimal(OP_READ, rid, p);
  if (nread < 0) {
    const errorTab: {[key: number]: string} = {
      10054: "Connection reset by peer.",
      10058: "Cannot send after socket shutdown."
    }
    console.log(nread*-1)
    throw new Error("read error " + (errorTab[(nread*-1)] || ""));
  } else if (nread == 0) {
    return { nread, eof: true };
  } else {
    return { nread, eof: false };
  }
}
```

**Comment by sasurau4** [comment]
@NamPNQ Thanks for your advice! I'll try them on this weekend!

**Comment by ry** [comment]
@sasurau4 Thanks for trying this but I guess it hasn't gone green, so we can't land. I'm closing because it's stale.

**Pull request closed without merging**